In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df_retail_data = pd.read_csv("Group5 - retail_data/retail_data_W23 - retail_data_W23.csv")
df_retail_order = pd.read_excel("Group5 - retail_data/retail_orders_W23.xlsx")
df_retail_store = pd.read_excel("Group5 - retail_data/store.xlsx")


In [3]:
df_retail_data

Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           4          5  2015-07-31       1498     1      1            0   
2           5          5  2015-07-31        559     1      1            0   
3           6          5  2015-07-31        589     1      1            0   
4          10          5  2015-07-31        681     1      1            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008   1106          2  2013-01-01          0     0      0            a   
651009   1109          2  2013-01-01          0     0      0            a   
651010   1112          2  2013-01-01          0     0      0            a   
651011   1113          2  2013-01-01          0     0      0            a   
651012   1114          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id  
0                   1       409  
1                   1       427  
2                   1       445  
3                   1       463  
4                   1       535  
...               ...       ...  
651008              1  18309955  
651009              1  18310009  
651010              1  18310063  
651011              1  18310081  
651012              1  18310099  

[651013 rows x 9 columns]

In [4]:
df_retail_order

Id  Orders
0            409    8314
1            427   13995
2            445    4822
3            463    5651
4            535    7185
...          ...     ...
651008  18309955       0
651009  18310009       0
651010  18310063       0
651011  18310081       0
651012  18310099       0

[651013 rows x 2 columns]

In [5]:
df_retail_store

Store StoreType Assortment  CompetitionDistance  \
0         1         c          a               1270.0   
1         2         a          a                570.0   
2         3         a          a              14130.0   
3         4         c          c                620.0   
4         5         a          a              29910.0   
...     ...       ...        ...                  ...   
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                           9.0                    2008.0       0   
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
3                           9.0                    2009.0       0   
4                           4.0                    2015.0       0   
...                         ...                       ...     ...   
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        NaN                       NaN       0   
1113                        NaN                       NaN       0   
1114                        NaN                       NaN       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  
0                 NaN              NaN               NaN  
1                13.0           2010.0   Jan,Apr,Jul,Oct  
2                14.0           2011.0   Jan,Apr,Jul,Oct  
3                 NaN              NaN               NaN  
4                 NaN              NaN               NaN  
...               ...              ...               ...  
1110             31.0           2013.0   Jan,Apr,Jul,Oct  
1111              NaN              NaN               NaN  
1112              NaN              NaN               NaN  
1113              NaN              NaN               NaN  
1114             22.0           2012.0  Mar,Jun,Sept,Dec  

[1115 rows x 10 columns]

In [6]:
for i in range(3):
    print(df_retail_order.describe())
    print(df_retail_order.describe())
    print(df_retail_store.describe())

                 Id         Orders
count  6.510130e+05  651013.000000
mean   9.153619e+06    5770.521006
std    5.285944e+06    3851.171979
min    4.090000e+02       0.000000
25%    4.578709e+06    3721.000000
50%    9.151915e+06    5743.000000
75%    1.373353e+07    7856.000000
max    1.831010e+07   38722.000000
                 Id         Orders
count  6.510130e+05  651013.000000
mean   9.153619e+06    5770.521006
std    5.285944e+06    3851.171979
min    4.090000e+02       0.000000
25%    4.578709e+06    3721.000000
50%    9.151915e+06    5743.000000
75%    1.373353e+07    7856.000000
max    1.831010e+07   38722.000000
            Store  CompetitionDistance  CompetitionOpenSinceMonth  \
count  1115.00000          1112.000000                 761.000000   
mean    558.00000          5404.901079                   7.224704   
std     322.01708          7663.174720                   3.212348   
min       1.00000            20.000000                   1.000000   
25%     279.50000        

In [7]:
df_retail_data['Id'] == df_retail_order['Id']

0         True
1         True
2         True
3         True
4         True
          ... 
651008    True
651009    True
651010    True
651011    True
651012    True
Name: Id, Length: 651013, dtype: bool

In [8]:
df_retail_data['Orders'] = df_retail_order['Orders']

In [9]:
df_retail_data


Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           4          5  2015-07-31       1498     1      1            0   
2           5          5  2015-07-31        559     1      1            0   
3           6          5  2015-07-31        589     1      1            0   
4          10          5  2015-07-31        681     1      1            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008   1106          2  2013-01-01          0     0      0            a   
651009   1109          2  2013-01-01          0     0      0            a   
651010   1112          2  2013-01-01          0     0      0            a   
651011   1113          2  2013-01-01          0     0      0            a   
651012   1114          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id  Orders  
0                   1       409    8314  
1                   1       427   13995  
2                   1       445    4822  
3                   1       463    5651  
4                   1       535    7185  
...               ...       ...     ...  
651008              1  18309955       0  
651009              1  18310009       0  
651010              1  18310063       0  
651011              1  18310081       0  
651012              1  18310099       0  

[651013 rows x 10 columns]

In [10]:
df_retail_data.isnull().sum()


Store            0
DayOfWeek        0
Date             0
Customers        0
Open             0
Promo            0
StateHoliday     0
SchoolHoliday    0
Id               0
Orders           0
dtype: int64

In [11]:
df_retail_data.nunique(axis=0)

Store              1115
DayOfWeek             7
Date                942
Customers          3915
Open                  2
Promo                 2
StateHoliday          4
SchoolHoliday         2
Id               651013
Orders            20206
dtype: int64

In [12]:
df_retail_store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [13]:
df_store_data = df_retail_store.dropna()

In [14]:
df_store_data.isnull().sum()

Store                        0
StoreType                    0
Assortment                   0
CompetitionDistance          0
CompetitionOpenSinceMonth    0
CompetitionOpenSinceYear     0
Promo2                       0
Promo2SinceWeek              0
Promo2SinceYear              0
PromoInterval                0
dtype: int64

In [15]:
df_store_data

Store StoreType Assortment  CompetitionDistance  \
1         2         a          a                570.0   
2         3         a          a              14130.0   
10       11         a          c                960.0   
13       14         a          a               1300.0   
14       15         d          c               4110.0   
...     ...       ...        ...                  ...   
1104   1105         c          c                330.0   
1105   1106         a          c               5330.0   
1106   1107         a          a               1400.0   
1108   1109         c          a               3490.0   
1110   1111         a          a               1900.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
10                         11.0                    2011.0       1   
13                          3.0                    2014.0       1   
14                          3.0                    2010.0       1   
...                         ...                       ...     ...   
1104                       11.0                    2008.0       1   
1105                        9.0                    2011.0       1   
1106                        6.0                    2012.0       1   
1108                        4.0                    2011.0       1   
1110                        6.0                    2014.0       1   

      Promo2SinceWeek  Promo2SinceYear    PromoInterval  
1                13.0           2010.0  Jan,Apr,Jul,Oct  
2                14.0           2011.0  Jan,Apr,Jul,Oct  
10                1.0           2012.0  Jan,Apr,Jul,Oct  
13               40.0           2011.0  Jan,Apr,Jul,Oct  
14               14.0           2011.0  Jan,Apr,Jul,Oct  
...               ...              ...              ...  
1104              5.0           2013.0  Feb,May,Aug,Nov  
1105             31.0           2013.0  Jan,Apr,Jul,Oct  
1106             13.0           2010.0  Jan,Apr,Jul,Oct  
1108             22.0           2012.0  Jan,Apr,Jul,Oct  
1110             31.0           2013.0  Jan,Apr,Jul,Oct  

[365 rows x 10 columns]

In [16]:
df_retail_and_store = pd.merge(df_retail_data, df_store_data, on='Store', how= 'outer')

In [17]:
df_retail_and_store

Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           3          4  2015-07-30        823     1      1            0   
2           3          2  2015-07-28        818     1      1            0   
3           3          7  2015-07-26          0     0      0            0   
4           3          6  2015-07-25        412     1      0            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008      1          1  2013-01-07        785     1      1            0   
651009      1          7  2013-01-06          0     0      0            0   
651010      1          6  2013-01-05        635     1      0            0   
651011      1          5  2013-01-04        619     1      0            0   
651012      1          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id  Orders StoreType Assortment  \
0                   1       409    8314         a          a   
1                   1     20479    8977         a          a   
2                   1     60619    8864         a          a   
3                   0    100759       0         a          a   
4                   0    120829    3878         a          a   
...               ...       ...     ...       ...        ...   
651008              1  18169663    7176       NaN        NaN   
651009              1  18189733       0       NaN        NaN   
651010              1  18209803    4997       NaN        NaN   
651011              1  18229873    4486       NaN        NaN   
651012              1  18290083       0       NaN        NaN   

        CompetitionDistance  CompetitionOpenSinceMonth  \
0                   14130.0                       12.0   
1                   14130.0                       12.0   
2                   14130.0                       12.0   
3                   14130.0                       12.0   
4                   14130.0                       12.0   
...                     ...                        ...   
651008                  NaN                        NaN   
651009                  NaN                        NaN   
651010                  NaN                        NaN   
651011                  NaN                        NaN   
651012                  NaN                        NaN   

        CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                         2006.0     1.0             14.0           2011.0   
1                         2006.0     1.0             14.0           2011.0   
2                         2006.0     1.0             14.0           2011.0   
3                         2006.0     1.0             14.0           2011.0   
4                         2006.0     1.0             14.0           2011.0   
...                          ...     ...              ...              ...   
651008                       NaN     NaN              NaN              NaN   
651009                       NaN     NaN              NaN              NaN   
651010                       NaN     NaN              NaN              NaN   
651011                       NaN     NaN              NaN              NaN   
651012                       NaN     NaN              NaN              NaN   

          PromoInterval  
0       Jan,Apr,Jul,Oct  
1       Jan,Apr,Jul,Oct  
2       Jan,Apr,Jul,Oct  
3       Jan,Apr,Jul,Oct  
4       Jan,Apr,Jul,Oct  
...                 ...  
651008              NaN  
651009              NaN  
651010              NaN  
651011              NaN  
651012              NaN  

[651013 rows x 19 columns]

In [18]:
df_retail_store_order = pd.merge(df_retail_and_store, df_retail_order, on='Id', how= 'outer')

In [19]:
df_retail_store_order


Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           3          4  2015-07-30        823     1      1            0   
2           3          2  2015-07-28        818     1      1            0   
3           3          7  2015-07-26          0     0      0            0   
4           3          6  2015-07-25        412     1      0            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008      1          1  2013-01-07        785     1      1            0   
651009      1          7  2013-01-06          0     0      0            0   
651010      1          6  2013-01-05        635     1      0            0   
651011      1          5  2013-01-04        619     1      0            0   
651012      1          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id  Orders_x StoreType Assortment  \
0                   1       409      8314         a          a   
1                   1     20479      8977         a          a   
2                   1     60619      8864         a          a   
3                   0    100759         0         a          a   
4                   0    120829      3878         a          a   
...               ...       ...       ...       ...        ...   
651008              1  18169663      7176       NaN        NaN   
651009              1  18189733         0       NaN        NaN   
651010              1  18209803      4997       NaN        NaN   
651011              1  18229873      4486       NaN        NaN   
651012              1  18290083         0       NaN        NaN   

        CompetitionDistance  CompetitionOpenSinceMonth  \
0                   14130.0                       12.0   
1                   14130.0                       12.0   
2                   14130.0                       12.0   
3                   14130.0                       12.0   
4                   14130.0                       12.0   
...                     ...                        ...   
651008                  NaN                        NaN   
651009                  NaN                        NaN   
651010                  NaN                        NaN   
651011                  NaN                        NaN   
651012                  NaN                        NaN   

        CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                         2006.0     1.0             14.0           2011.0   
1                         2006.0     1.0             14.0           2011.0   
2                         2006.0     1.0             14.0           2011.0   
3                         2006.0     1.0             14.0           2011.0   
4                         2006.0     1.0             14.0           2011.0   
...                          ...     ...              ...              ...   
651008                       NaN     NaN              NaN              NaN   
651009                       NaN     NaN              NaN              NaN   
651010                       NaN     NaN              NaN              NaN   
651011                       NaN     NaN              NaN              NaN   
651012                       NaN     NaN              NaN              NaN   

          PromoInterval  Orders_y  
0       Jan,Apr,Jul,Oct      8314  
1       Jan,Apr,Jul,Oct      8977  
2       Jan,Apr,Jul,Oct      8864  
3       Jan,Apr,Jul,Oct         0  
4       Jan,Apr,Jul,Oct      3878  
...                 ...       ...  
651008              NaN      7176  
651009              NaN         0  
651010              NaN      4997  
651011              NaN      4486  
651012              NaN         0  

[651013 rows x 20 columns]

In [20]:
df_retail_store_order.describe()

Store      DayOfWeek      Customers           Open  \
count  651013.000000  651013.000000  651013.000000  651013.000000   
mean      558.645629       3.999336     632.851832       0.829619   
std       321.905872       1.998260     464.857658       0.375967   
min         1.000000       1.000000       0.000000       0.000000   
25%       281.000000       2.000000     404.000000       1.000000   
50%       558.000000       4.000000     609.000000       1.000000   
75%       838.000000       6.000000     837.000000       1.000000   
max      1115.000000       7.000000    5458.000000       1.000000   

               Promo  SchoolHoliday            Id       Orders_x  \
count  651013.000000  651013.000000  6.510130e+05  651013.000000   
mean        0.381558       0.178927  9.153619e+06    5770.521006   
std         0.485769       0.383292  5.285944e+06    3851.171979   
min         0.000000       0.000000  4.090000e+02       0.000000   
25%         0.000000       0.000000  4.578709e+06    3721.000000   
50%         0.000000       0.000000  9.151915e+06    5743.000000   
75%         1.000000       0.000000  1.373353e+07    7856.000000   
max         1.000000       1.000000  1.831010e+07   38722.000000   

       CompetitionDistance  CompetitionOpenSinceMonth  \
count        207314.000000              207314.000000   
mean           4294.989533                   7.192384   
std            5260.803188                   3.248784   
min              40.000000                   1.000000   
25%             600.000000                   4.000000   
50%            2240.000000                   7.000000   
75%            5630.000000                  10.000000   
max           27190.000000                  12.000000   

       CompetitionOpenSinceYear    Promo2  Promo2SinceWeek  Promo2SinceYear  \
count             207314.000000  207314.0    207314.000000    207314.000000   
mean                2008.171990       1.0        22.132774      2011.854655   
std                    7.513637       0.0        13.540669         1.674349   
min                 1900.000000       1.0         1.000000      2009.000000   
25%                 2005.000000       1.0        13.000000      2011.000000   
50%                 2009.000000       1.0        18.000000      2012.000000   
75%                 2012.000000       1.0        37.000000      2013.000000   
max                 2015.000000       1.0        48.000000      2015.000000   

            Orders_y  
count  651013.000000  
mean     5770.521006  
std      3851.171979  
min         0.000000  
25%      3721.000000  
50%      5743.000000  
75%      7856.000000  
max     38722.000000

In [23]:
df_retail_store_order.columns

Index(['Store', 'DayOfWeek', 'Date', 'Customers', 'Open', 'Promo',
       'StateHoliday', 'SchoolHoliday', 'Id', 'Orders_x', 'StoreType',
       'Assortment', 'CompetitionDistance', 'CompetitionOpenSinceMonth',
       'CompetitionOpenSinceYear', 'Promo2', 'Promo2SinceWeek',
       'Promo2SinceYear', 'PromoInterval', 'Orders_y'],
      dtype='object')

In [24]:
columns_are_equal = df_retail_store_order['Orders_x'].equals(df_retail_store_order['Orders_y'])
print(columns_are_equal)

True


In [25]:
df_retail_store_order = df_retail_store_order.drop('Orders_x', axis=1)

In [26]:
df_retail_store_order

Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           3          4  2015-07-30        823     1      1            0   
2           3          2  2015-07-28        818     1      1            0   
3           3          7  2015-07-26          0     0      0            0   
4           3          6  2015-07-25        412     1      0            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008      1          1  2013-01-07        785     1      1            0   
651009      1          7  2013-01-06          0     0      0            0   
651010      1          6  2013-01-05        635     1      0            0   
651011      1          5  2013-01-04        619     1      0            0   
651012      1          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id StoreType Assortment  CompetitionDistance  \
0                   1       409         a          a              14130.0   
1                   1     20479         a          a              14130.0   
2                   1     60619         a          a              14130.0   
3                   0    100759         a          a              14130.0   
4                   0    120829         a          a              14130.0   
...               ...       ...       ...        ...                  ...   
651008              1  18169663       NaN        NaN                  NaN   
651009              1  18189733       NaN        NaN                  NaN   
651010              1  18209803       NaN        NaN                  NaN   
651011              1  18229873       NaN        NaN                  NaN   
651012              1  18290083       NaN        NaN                  NaN   

        CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                            12.0                    2006.0     1.0   
1                            12.0                    2006.0     1.0   
2                            12.0                    2006.0     1.0   
3                            12.0                    2006.0     1.0   
4                            12.0                    2006.0     1.0   
...                           ...                       ...     ...   
651008                        NaN                       NaN     NaN   
651009                        NaN                       NaN     NaN   
651010                        NaN                       NaN     NaN   
651011                        NaN                       NaN     NaN   
651012                        NaN                       NaN     NaN   

        Promo2SinceWeek  Promo2SinceYear    PromoInterval  Orders_y  
0                  14.0           2011.0  Jan,Apr,Jul,Oct      8314  
1                  14.0           2011.0  Jan,Apr,Jul,Oct      8977  
2                  14.0           2011.0  Jan,Apr,Jul,Oct      8864  
3                  14.0           2011.0  Jan,Apr,Jul,Oct         0  
4                  14.0           2011.0  Jan,Apr,Jul,Oct      3878  
...                 ...              ...              ...       ...  
651008              NaN              NaN              NaN      7176  
651009              NaN              NaN              NaN         0  
651010              NaN              NaN              NaN      4997  
651011              NaN              NaN              NaN      4486  
651012              NaN              NaN              NaN         0  

[651013 rows x 19 columns]

In [27]:
duplicate_columns = df_retail_store_order.T.duplicated()

# Step 2: Get the names of the removed columns
removed_columns = df_retail_store_order.columns[duplicate_columns]

# Step 3: Drop the duplicate columns
df_retail_store_order = df_retail_store_order.loc[:, ~duplicate_columns]

print("DataFrame after removing duplicate columns:")
print(df_retail_store_order)

print("Names of removed columns:")
print(removed_columns)

DataFrame after removing duplicate columns:
        Store  DayOfWeek        Date  Customers  Open  Promo StateHoliday  \
0           3          5  2015-07-31        821     1      1            0   
1           3          4  2015-07-30        823     1      1            0   
2           3          2  2015-07-28        818     1      1            0   
3           3          7  2015-07-26          0     0      0            0   
4           3          6  2015-07-25        412     1      0            0   
...       ...        ...         ...        ...   ...    ...          ...   
651008      1          1  2013-01-07        785     1      1            0   
651009      1          7  2013-01-06          0     0      0            0   
651010      1          6  2013-01-05        635     1      0            0   
651011      1          5  2013-01-04        619     1      0            0   
651012      1          2  2013-01-01          0     0      0            a   

        SchoolHoliday        Id

In [28]:
removed_columns

Index([], dtype='object')

In [30]:
df_retail_store_order.isnull().sum()

Store                             0
DayOfWeek                         0
Date                              0
Customers                         0
Open                              0
Promo                             0
StateHoliday                      0
SchoolHoliday                     0
Id                                0
StoreType                    443699
Assortment                   443699
CompetitionDistance          443699
CompetitionOpenSinceMonth    443699
CompetitionOpenSinceYear     443699
Promo2                       443699
Promo2SinceWeek              443699
Promo2SinceYear              443699
PromoInterval                443699
Orders_y                          0
dtype: int64

In [31]:
df_retail_store_order = df_retail_store_order.dropna()

In [32]:
df_retail_store_order.isnull().sum()

Store                        0
DayOfWeek                    0
Date                         0
Customers                    0
Open                         0
Promo                        0
StateHoliday                 0
SchoolHoliday                0
Id                           0
StoreType                    0
Assortment                   0
CompetitionDistance          0
CompetitionOpenSinceMonth    0
CompetitionOpenSinceYear     0
Promo2                       0
Promo2SinceWeek              0
Promo2SinceYear              0
PromoInterval                0
Orders_y                     0
dtype: int64

In [33]:
df_retail_store_order.shape

(207314, 19)

In [35]:
column_data_types = df_retail_store_order.dtypes

print("Data Types of each column:")
print(column_data_types)


Data Types of each column:
Store                          int64
DayOfWeek                      int64
Date                          object
Customers                      int64
Open                           int64
Promo                          int64
StateHoliday                  object
SchoolHoliday                  int64
Id                             int64
StoreType                     object
Assortment                    object
CompetitionDistance          float64
CompetitionOpenSinceMonth    float64
CompetitionOpenSinceYear     float64
Promo2                       float64
Promo2SinceWeek              float64
Promo2SinceYear              float64
PromoInterval                 object
Orders_y                       int64
dtype: object


In [37]:
object_columns = df_retail_store_order.dtypes[df_retail_store_order.dtypes == 'object']
print("Columns with object data type:")
print(object_columns)

Columns with object data type:
Date             object
StateHoliday     object
StoreType        object
Assortment       object
PromoInterval    object
dtype: object


In [49]:
df_retail_store_order["PromoInterval"].unique()

array(['Jan,Apr,Jul,Oct', 'Mar,Jun,Sept,Dec', 'Feb,May,Aug,Nov'],
      dtype=object)

In [48]:
df_retail_store_order["Assortment"].unique()

array(['a', 'c'], dtype=object)

In [47]:
df_retail_store_order["StoreType"].unique()

array(['a', 'd', 'c'], dtype=object)

In [44]:
df_retail_store_order["StateHoliday"].unique()

array(['0', 'a', 'b', 'c'], dtype=object)

In [40]:
df_retail_store_order["Date"]

0         2015-07-31
1         2015-07-30
2         2015-07-28
3         2015-07-26
4         2015-07-25
             ...    
649205    2013-01-06
649206    2013-01-05
649207    2013-01-04
649208    2013-01-02
649209    2013-01-01
Name: Date, Length: 207314, dtype: object

In [55]:

# Convert the 'Date' column to a datetime format if not already in that format
df_retail_store_order['Date'] = pd.to_datetime(df_retail_store_order['Date'])

# Convert the 'Date' column to a timestamp (Unix timestamp in seconds)
df_retail_store_order['Date_stamp'] = (df_retail_store_order['Date'] - pd.Timestamp("1970-01-01")) // pd.Timedelta(seconds=1)

print("DataFrame after converting Date to timestamp:")
print(df_retail_store_order)


DataFrame after converting Date to timestamp:
        Store  DayOfWeek       Date  Customers  Open  Promo StateHoliday  \
0           3          5 2015-07-31        821     1      1            0   
1           3          4 2015-07-30        823     1      1            0   
2           3          2 2015-07-28        818     1      1            0   
3           3          7 2015-07-26          0     0      0            0   
4           3          6 2015-07-25        412     1      0            0   
...       ...        ...        ...        ...   ...    ...          ...   
649205   1026          7 2013-01-06          0     0      0            0   
649206   1026          6 2013-01-05       1051     1      0            0   
649207   1026          5 2013-01-04       1267     1      0            0   
649208   1026          3 2013-01-02       1237     1      0            0   
649209   1026          2 2013-01-01          0     0      0            a   

        SchoolHoliday        Id StoreType

In [56]:
# Now drop the original 'Date' column as we have extracted its components
df_retail_store_order.drop('Date', axis=1, inplace=True)

print("DataFrame after extracting date components as features:")
print(df_retail_store_order)

DataFrame after extracting date components as features:
        Store  DayOfWeek  Customers  Open  Promo StateHoliday  SchoolHoliday  \
0           3          5        821     1      1            0              1   
1           3          4        823     1      1            0              1   
2           3          2        818     1      1            0              1   
3           3          7          0     0      0            0              0   
4           3          6        412     1      0            0              0   
...       ...        ...        ...   ...    ...          ...            ...   
649205   1026          7          0     0      0            0              0   
649206   1026          6       1051     1      0            0              0   
649207   1026          5       1267     1      0            0              1   
649208   1026          3       1237     1      0            0              1   
649209   1026          2          0     0      0            a   

In [57]:



# Assuming 'PromoInterval' column contains values like "Jan,Apr,Jul,Oct"

# Split the 'PromoInterval' column into separate columns for each month
months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec']
for month in months:
    df_retail_store_order[month] = df_retail_store_order['PromoInterval'].str.contains(month).astype(int)

# Now drop the original 'PromoInterval' column as we have extracted its information
df_retail_store_order.drop('PromoInterval', axis=1, inplace=True)

print("DataFrame after preprocessing PromoInterval:")
print(df_retail_store_order)


DataFrame after preprocessing PromoInterval:
        Store  DayOfWeek  Customers  Open  Promo StateHoliday  SchoolHoliday  \
0           3          5        821     1      1            0              1   
1           3          4        823     1      1            0              1   
2           3          2        818     1      1            0              1   
3           3          7          0     0      0            0              0   
4           3          6        412     1      0            0              0   
...       ...        ...        ...   ...    ...          ...            ...   
649205   1026          7          0     0      0            0              0   
649206   1026          6       1051     1      0            0              0   
649207   1026          5       1267     1      0            0              1   
649208   1026          3       1237     1      0            0              1   
649209   1026          2          0     0      0            a              

In [60]:
object_columns = df_retail_store_order.dtypes[df_retail_store_order.dtypes == 'object']
print("Columns with object data type:")
print(object_columns)

Columns with object data type:
StateHoliday    object
StoreType       object
Assortment      object
dtype: object


In [61]:
import pandas as pd

# Assuming df is your DataFrame containing the columns 'StateHoliday', 'StoreType', and 'Assortment'

# Perform one-hot encoding on the 'StateHoliday' column
df_retail_store_order_encoded = pd.get_dummies(df_retail_store_order, columns=['StateHoliday'], prefix='StateHoliday')

# Perform one-hot encoding on the 'StoreType' column
df_retail_store_order_encoded = pd.get_dummies(df_retail_store_order_encoded, columns=['StoreType'], prefix='StoreType')

# Perform one-hot encoding on the 'Assortment' column
df_retail_store_order_encoded = pd.get_dummies(df_retail_store_order_encoded, columns=['Assortment'], prefix='Assortment')

print("DataFrame after one-hot encoding:")
print(df_retail_store_order_encoded)


DataFrame after one-hot encoding:
        Store  DayOfWeek  Customers  Open  Promo  SchoolHoliday        Id  \
0           3          5        821     1      1              1       409   
1           3          4        823     1      1              1     20479   
2           3          2        818     1      1              1     60619   
3           3          7          0     0      0              0    100759   
4           3          6        412     1      0              0    120829   
...       ...        ...        ...   ...    ...            ...       ...   
649205   1026          7          0     0      0              0  18208183   
649206   1026          6       1051     1      0              0  18228253   
649207   1026          5       1267     1      0              1  18248323   
649208   1026          3       1237     1      0              1  18288463   
649209   1026          2          0     0      0              1  18308515   

        CompetitionDistance  CompetitionO

In [62]:
object_columns = df_retail_store_order_encoded.dtypes[df_retail_store_order_encoded.dtypes == 'object']
print("Columns with object data type:")
print(object_columns)

Columns with object data type:
Series([], dtype: object)


In [65]:
object_columns = df_retail_store_order_encoded.dtypes
print("Columns with object data type:")
print(object_columns)

Columns with object data type:
Store                          int64
DayOfWeek                      int64
Customers                      int64
Open                           int64
Promo                          int64
SchoolHoliday                  int64
Id                             int64
CompetitionDistance          float64
CompetitionOpenSinceMonth    float64
CompetitionOpenSinceYear     float64
Promo2                       float64
Promo2SinceWeek              float64
Promo2SinceYear              float64
Orders_y                       int64
Date_stamp                     int64
Jan                            int32
Feb                            int32
Mar                            int32
Apr                            int32
May                            int32
Jun                            int32
Jul                            int32
Aug                            int32
Sept                           int32
Oct                            int32
Nov                            int32
Dec    

In [66]:
df_retail_store_order_encoded.columns

Index(['Store', 'DayOfWeek', 'Customers', 'Open', 'Promo', 'SchoolHoliday',
       'Id', 'CompetitionDistance', 'CompetitionOpenSinceMonth',
       'CompetitionOpenSinceYear', 'Promo2', 'Promo2SinceWeek',
       'Promo2SinceYear', 'Orders_y', 'Date_stamp', 'Jan', 'Feb', 'Mar', 'Apr',
       'May', 'Jun', 'Jul', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec',
       'StateHoliday_0', 'StateHoliday_a', 'StateHoliday_b', 'StateHoliday_c',
       'StoreType_a', 'StoreType_c', 'StoreType_d', 'Assortment_a',
       'Assortment_c'],
      dtype='object')

In [67]:
X = df_retail_store_order_encoded.drop('Orders_y', axis=1)
y = df_retail_store_order_encoded['Orders_y']

In [68]:
X

Store  DayOfWeek  Customers  Open  Promo  SchoolHoliday        Id  \
0           3          5        821     1      1              1       409   
1           3          4        823     1      1              1     20479   
2           3          2        818     1      1              1     60619   
3           3          7          0     0      0              0    100759   
4           3          6        412     1      0              0    120829   
...       ...        ...        ...   ...    ...            ...       ...   
649205   1026          7          0     0      0              0  18208183   
649206   1026          6       1051     1      0              0  18228253   
649207   1026          5       1267     1      0              1  18248323   
649208   1026          3       1237     1      0              1  18288463   
649209   1026          2          0     0      0              1  18308515   

        CompetitionDistance  CompetitionOpenSinceMonth  \
0                   14130.0                       12.0   
1                   14130.0                       12.0   
2                   14130.0                       12.0   
3                   14130.0                       12.0   
4                   14130.0                       12.0   
...                     ...                        ...   
649205                450.0                        6.0   
649206                450.0                        6.0   
649207                450.0                        6.0   
649208                450.0                        6.0   
649209                450.0                        6.0   

        CompetitionOpenSinceYear  ...  Dec  StateHoliday_0  StateHoliday_a  \
0                         2006.0  ...    0               1               0   
1                         2006.0  ...    0               1               0   
2                         2006.0  ...    0               1               0   
3                         2006.0  ...    0               1               0   
4                         2006.0  ...    0               1               0   
...                          ...  ...  ...             ...             ...   
649205                    2011.0  ...    1               1               0   
649206                    2011.0  ...    1               1               0   
649207                    2011.0  ...    1               1               0   
649208                    2011.0  ...    1               1               0   
649209                    2011.0  ...    1               0               1   

        StateHoliday_b  StateHoliday_c  StoreType_a  StoreType_c  StoreType_d  \
0                    0               0            1            0            0   
1                    0               0            1            0            0   
2                    0               0            1            0            0   
3                    0               0            1            0            0   
4                    0               0            1            0            0   
...                ...             ...          ...          ...          ...   
649205               0               0            0            1            0   
649206               0               0            0            1            0   
649207               0               0            0            1            0   
649208               0               0            0            1            0   
649209               0               0            0            1            0   

        Assortment_a  Assortment_c  
0                  1             0  
1                  1             0  
2                  1             0  
3                  1             0  
4                  1             0  
...              ...           ...  
649205             1             0  
649206             1             0  
649207             1             0  
649208             1             0  
649209             1             0  

[207314 rows x 35 columns]

In [69]:
y

0         8314
1         8977
2         8864
3            0
4         3878
          ... 
649205       0
649206    7776
649207    8619
649208    8320
649209       0
Name: Orders_y, Length: 207314, dtype: int64

In [70]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow import keras
from tensorflow.keras import layers

# Load your dataset (X contains features, y contains target labels)
# Assuming X and y are your feature matrix and target vector, respectively

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Standardize the features (optional but often helpful for DNN)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)



#### XGboost

In [75]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import xgboost as xgb

# Load your dataset (X contains features, y contains target labels)
# Assuming X and y are your feature matrix and target vector, respectively

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Convert data to DMatrix format (required by XGBoost)
dtrain = xgb.DMatrix(data=X_train, label=y_train)
dtest = xgb.DMatrix(data=X_test, label=y_test)

# Define XGBoost parameters for regression
params = {
    'objective': 'reg:squarederror',  # For regression tasks, use 'reg:squarederror'
    'eval_metric': 'rmse',            # Root Mean Squared Error (RMSE) for evaluation
    'max_depth': 3,                   # Maximum depth of the trees
    'eta': 0.1,                       # Learning rate
    'subsample': 0.8,                 # Subsample ratio of the training instances
    'colsample_bytree': 0.8           # Subsample ratio of columns when constructing each tree
}

# Train the XGBoost model for regression
num_rounds = 100  # Number of boosting rounds (iterations)
model = xgb.train(params, dtrain, num_rounds)

# Make predictions on the test set
y_pred = model.predict(dtest)

# Evaluate the model using appropriate regression metrics (optional)
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print("Mean Squared Error (MSE):", mse)
print("Mean Absolute Error (MAE):", mae)
print("R-squared (R2) Score:", r2)


Mean Squared Error (MSE): 640686.5808925513
Mean Absolute Error (MAE): 561.7360434159153
R-squared (R2) Score: 0.9472773027561886


In [77]:
import numpy as np
import pandas as pd
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split
import xgboost as xgb

# Load your dataset (X contains features, y contains target labels)
# Assuming X and y are your feature matrix and target vector, respectively

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the XGBoost model
model = xgb.XGBRegressor()

# Perform Recursive Feature Elimination (RFE) to select top features
n_features_to_select = 10  # Specify the number of top features you want to keep
rfe = RFE(estimator=model, n_features_to_select=n_features_to_select)
X_train_selected = rfe.fit_transform(X_train, y_train)
X_test_selected = rfe.transform(X_test)

# Convert data to DMatrix format (required by XGBoost)
dtrain = xgb.DMatrix(data=X_train_selected, label=y_train)
dtest = xgb.DMatrix(data=X_test_selected, label=y_test)

# Define XGBoost parameters for regression
params = {
    'objective': 'reg:squarederror',  # For regression tasks, use 'reg:squarederror'
    'eval_metric': 'rmse',            # Root Mean Squared Error (RMSE) for evaluation
    'max_depth': 3,                   # Maximum depth of the trees
    'eta': 0.1,                       # Learning rate
    'subsample': 0.8,                 # Subsample ratio of the training instances
    'colsample_bytree': 0.8           # Subsample ratio of columns when constructing each tree
}

# Train the XGBoost model for regression
num_rounds = 100  # Number of boosting rounds (iterations)
model = xgb.train(params, dtrain, num_rounds)

# Make predictions on the test set
y_pred = model.predict(dtest)

# Evaluate the model using appropriate regression metrics (optional)
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print("Mean Squared Error (MSE):", mse)
print("Mean Absolute Error (MAE):", mae)
print("R-squared (R2) Score:", r2)


Mean Squared Error (MSE): 730409.455765927
Mean Absolute Error (MAE): 607.8002742305708
R-squared (R2) Score: 0.9398939235675636


In [78]:
y_pred

array([4719.031  ,   80.33672, 7007.625  , ..., 5991.489  , 6687.857  ,
       7022.005  ], dtype=float32)

In [84]:
# Compute the Mean Percentage Error (MPE) with handling for zero division
epsilon = 1e-8  # A small epsilon value to avoid division by zero
mpe = np.mean((y_test - y_pred) / (y_test + epsilon)) * 100
print("Mean Percentage Error (MPE):", mpe)


Mean Percentage Error (MPE): -51943367944.90301


In [85]:
mbd = np.mean(y_test - y_pred)
print("Mean Bias Deviation (MBD):", mbd)

Mean Bias Deviation (MBD): -1.4756449442798418


In [86]:
from sklearn.metrics import r2_score
n = X_test.shape[0]  # Number of samples
p = X_test.shape[1]  # Number of features
cod = 1 - (1 - r2) * (n - 1) / (n - p - 1)
print("Coefficient of Determination (COD):", cod)


Coefficient of Determination (COD): 0.9398431423699115


An R2 score of 0.9398431423699115 is relatively high, close to 1, which suggests that your model explains about 93.98% of the variance in the target variable. In general, a high R2 score indicates that the model's predictions are closely fitting the actual data, and it is considered a good result.

In [87]:
from sklearn.metrics import explained_variance_score
ev = explained_variance_score(y_test, y_pred)
print("Explained Variance Score (EV):", ev)


Explained Variance Score (EV): 0.939894102758377


An Explained Variance Score of 0.939894102758377 is again quite high, close to 1, which indicates that your model explains approximately 93.99% of the variance in the target variable. This is consistent with the high R2 score you obtained earlier.